pkg/sorter/pebble/event_sorter.go

```diff
@@ -0,0 +1,404 @@
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	s.onResolves = append(s.onResolves, action)
+}
+
+// FetchByTable implements sorter.EventSortEngine.
+func (s *EventSorter) FetchByTable(
+	tableID model.TableID,
+	lowerBound, upperBound sorter.Position,
+) sorter.EventIterator {
+	s.mu.RLock()
+	state, exists := s.tables[tableID]
+	s.mu.RUnlock()
+
+	if !exists {
+		log.Panic("fetch events from an unexist table",
+			zap.String("namespace", s.changefeedID.Namespace),
+			zap.String("changefeed", s.changefeedID.ID))
+	}
+
+	if upperBound.CommitTs > atomic.LoadUint64(&state.sortedResolved) {
+		log.Panic("fetch unresolved events",
+			zap.String("namespace", s.changefeedID.Namespace),
+			zap.String("changefeed", s.changefeedID.ID))
+	}
+
+	db := s.dbs[getDB(tableID, len(s.dbs))]
+	iter := iterTable(db, s.uniqueID, tableID, lowerBound, upperBound)
+	return &EventIter{tableID: tableID, iter: iter, serde: s.serde}
+}
+
+// FetchAllTables implements sorter.EventSortEngine.
+func (s *EventSorter) FetchAllTables(lowerBound sorter.Position) sorter.EventIterator {
+	log.Panic("FetchAllTables should never be called",
+		zap.String("namespace", s.changefeedID.Namespace),
+		zap.String("changefeed", s.changefeedID.ID))
+	return nil
+}
+
+// CleanByTable implements sorter.EventSortEngine.
+func (s *EventSorter) CleanByTable(tableID model.TableID, upperBound sorter.Position) error {
+	s.mu.RLock()
+	state, exists := s.tables[tableID]
+	s.mu.RUnlock()
+
+	if !exists {
+		log.Panic("clean an unexist table",
+			zap.String("namespace", s.changefeedID.Namespace),
+			zap.String("changefeed", s.changefeedID.ID))
+	}
+
+	return s.cleanTable(state, tableID, upperBound)
+}
+
+// CleanAllTables implements sorter.EventSortEngine.
+func (s *EventSorter) CleanAllTables(upperBound sorter.Position) error {
+	log.Panic("CleanAllTables should never be called",
+		zap.String("namespace", s.changefeedID.Namespace),
+		zap.String("changefeed", s.changefeedID.ID))
+	return nil
+}
+
+// Close implements sorter.EventSortEngine.
+func (s *EventSorter) Close() error {
+	close(s.closed)
+	s.wg.Wait()
+	for _, ch := range s.channs {
+		ch.Close()
+		for range ch.Out() {
+		}
+	}
+
+	s.mu.RLock()
+	defer s.mu.RUnlock()
+	for tableID, state := range s.tables {
+		if err := s.cleanTable(state, tableID); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// Next implements sorter.EventIterator.
+func (s *EventIter) Next() (event *model.PolymorphicEvent, pos sorter.Position, err error) {
+	var valid bool = s.iter != nil && s.iter.Valid()
+	var value []byte = nil
+	for valid {
+		value, valid = s.iter.Value(), s.iter.Next()
+		event = &model.PolymorphicEvent{}
+		if _, err = s.serde.Unmarshal(event, value); err != nil {
+			return
+		}
+		if s.headItem != nil {
+			break
+		}
+		s.headItem, event = event, nil
+	}
+	if s.headItem != nil {
+		if event == nil || s.headItem.CRTs != event.CRTs || s.headItem.StartTs != event.StartTs {
```

Comment:
if event != nil  & s.headItem.CRTs == event.CTRs & s.headItem.StartTs == event.StartTs ,  what is the value of pos?

Reply:
In this case `pos` won't be updated. `pos` will only be filled when meets a finished transaction.